core/server/proxy/src/main/java/alluxio/proxy/s3/CompleteMultipartUploadHandler.java

```diff
@@ -59,56 +62,80 @@ public class CompleteMultipartUploadHandler extends AbstractHandler {
     mFileSystem = fs;
     mExecutor = Executors.newFixedThreadPool(Configuration.getInt(
         PropertyKey.PROXY_S3_COMPLETE_MULTIPART_UPLOAD_POOL_SIZE));
     mKeepAliveTime = Configuration.getMs(
         PropertyKey.PROXY_S3_COMPLETE_MULTIPART_UPLOAD_KEEPALIVE_TIME_INTERVAL);
+    mS3Prefix = baseUri + AlluxioURI.SEPARATOR + S3RestServiceHandler.SERVICE_PREFIX
+        + AlluxioURI.SEPARATOR;
   }
 
   /**
    * Process s3 multipart upload request.
    */
   @Override
-  public void handle(String s, Request request, HttpServletRequest
-      httpServletRequest, HttpServletResponse httpServletResponse)
-      throws IOException, ServletException {
-    if (s.startsWith(mS3Prefix)
-        && request.getMethod().equals("POST")
-        && request.getParameter("uploadId") != null) {
-
-      final String bucket = s.substring(mS3Prefix.length(), s.lastIndexOf(AlluxioURI.SEPARATOR));
-      final String object = s.substring(s.lastIndexOf(AlluxioURI.SEPARATOR) + 1);
-      final Long uploadId = Long.valueOf(request.getParameter("uploadId"));
-      httpServletResponse.setStatus(HttpServletResponse.SC_OK);
-      httpServletResponse.setContentType("text/xml");
-
-      Future<CompleteMultipartUploadResult> future =
-          mExecutor.submit(new CompleteMultipartUploadTask(mFileSystem, bucket, object, uploadId));
-      while (!future.isDone()) {
-        try {
-          Thread.sleep(mKeepAliveTime);
-        } catch (InterruptedException e) {
-          e.printStackTrace();
-        }
-        //periodically sends white space characters to keep the connection from timing out
-        httpServletResponse.getWriter().print(" ");
-        httpServletResponse.getWriter().flush();
-      }
+  public void handle(String s, Request request, HttpServletRequest httpServletRequest,
+                     HttpServletResponse httpServletResponse) throws IOException {
+    if (!s.startsWith(mS3Prefix)) {
+      return;
+    }
+    LOG.info("Alluxio S3 API received request: {}", request);
+    if (!request.getMethod().equals("POST")
+        || request.getParameter("uploadId") == null) {
+      return;
+    } // Otherwise, handle CompleteMultipartUpload
+    s = s.substring(mS3Prefix.length());
+    final String bucket = s.substring(0, s.indexOf(AlluxioURI.SEPARATOR));
+    final String object = s.substring(s.indexOf(AlluxioURI.SEPARATOR) + 1);
+    final String uploadId = request.getParameter("uploadId");
+    LOG.debug("(bucket: {}, object: {}, uploadId: {}) queuing task...",
+        bucket, object, uploadId);
 
-      XmlMapper mapper = new XmlMapper();
+    httpServletResponse.setStatus(HttpServletResponse.SC_OK);
+    httpServletResponse.setContentType(MediaType.APPLICATION_XML);
 
+    Future<CompleteMultipartUploadResult> future =
+        mExecutor.submit(new CompleteMultipartUploadTask(mFileSystem, bucket, object, uploadId,
+            IOUtils.toString(request.getReader())));
+    long sleepMs = 1000;
+    while (!future.isDone()) {
```

Comment:
if we are blocked on the future anyways, we should just use future.get() to block?

Reply:
No because we need to be able to send whitespace characters to keep the connection alive. If we are blocked indefinitely on `future.get()` then we can exceed the timeout on the HTTP request before getting a response.